src/item.rs

```diff
@@ -1,33 +1,102 @@
+
 use bevy::{
+    ecs::system::EntityCommands,
+    hierarchy::{BuildChildren, Children},
     math::Vec3,
-    prelude::{Bundle, Handle, Transform},
+    prelude::{Assets, Bundle, Commands, Component, Entity, Handle, Query, Transform, With},
     transform::TransformBundle,
 };
+use leafwing_input_manager::prelude::ActionState;
 
 use crate::{
-    consts::{self, ITEM_LAYER},
+    consts::{self, ITEM_LAYER, PICK_ITEM_RADIUS},
+    input::PlayerAction,
     metadata::{ItemMeta, ItemSpawnMeta},
+    player::Player,
+    state::State,
 };
 
+#[derive(Component)]
+pub struct Item;
+
+/// Represents an item, that is either on the map (waiting to be picked up), or carried.
+/// If an item is on the map, it has a TransformBundle; if it's carried, it doesn't, and it's the
+/// child entity of a Player.
 #[derive(Bundle)]
-pub struct ItemSpawnBundle {
+pub struct ItemBundle {
+    item: Item,
     item_meta_handle: Handle<ItemMeta>,
-    #[bundle]
-    transform_bundle: TransformBundle,
 }
 
-impl ItemSpawnBundle {
+impl ItemBundle {
     pub fn new(item_spawn_meta: &ItemSpawnMeta) -> Self {
-        let item_meta_handle = item_spawn_meta.item_handle.clone();
+        Self {
+            item: Item,
+            item_meta_handle: item_spawn_meta.item_handle.clone(),
+        }
+    }
 
+    pub fn spawn(mut commands: EntityCommands, item_spawn_meta: &ItemSpawnMeta) {
         let ground_offset = Vec3::new(0.0, consts::GROUND_Y, ITEM_LAYER);
         let transform_bundle = TransformBundle::from_transform(Transform::from_translation(
             item_spawn_meta.location + ground_offset,
         ));
 
-        Self {
-            item_meta_handle,
-            transform_bundle,
+        commands.insert_bundle(transform_bundle);
+    }
+}
+
+pub fn pick_items(
+    mut commands: Commands,
+    player_query: Query<(Entity, &Transform, &State, &ActionState<PlayerAction>), With<Player>>,
+    items_query: Query<(Entity, &Transform), With<Item>>,
+) {
+    // We need to track the picked items, otherwise, in theory, two players could pick the same item.
+    let mut picked_item_ids = HashSet::new();
+
+    for (player_id, player_transform, player_state, input) in player_query.iter() {
+        if *player_state != State::Idle && *player_state != State::Running {
+            continue;
+        }
+
+        if input.just_pressed(PlayerAction::Shoot) || input.just_pressed(PlayerAction::Throw) {
+            // If several items are at pick distance, an arbitrary one is picked.
+            for (item_id, item_transform) in items_query.iter() {
+                if !picked_item_ids.contains(&item_id) {
+                    let player_item_distance = player_transform
+                        .translation
+                        .truncate()
+                        .distance(item_transform.translation.truncate());
+
+                    if player_item_distance <= PICK_ITEM_RADIUS {
+                        commands.entity(item_id).remove_bundle::<TransformBundle>();
```

Comment:
I meant just to keep the transform bundle either way but I guess this is fine as well, we can always change it later if we need to use the transform for animations in the future.

Reply:
ok. if we put it in the hads or something, then it will certainly go back :smile: it is unnecessary now, since if we put it in the HUD (there is a tracking issue), the HUD uses different semnantics/types anyway.